routines/Random.asm

```diff
@@ -3,32 +3,42 @@
 ; Output: A = random number in the interval [0, A]
  
 ?main:
-    PHX : PHP
-    SEP #$30
-    PHA
-    JSL $01ACF9|!BankB
-    LDA $148D|!Base2
-    PLX
-    CPX #$FF
-    BEQ ?.end
- 
-?.normal
-    INX
- 
-    if !SA1 == 0
-        STA $4202               ; Write first multiplicand.
-        STX $4203               ; Write second multiplicand.
-        NOP #4                  ; Wait 8 cycles.
-        LDA $4217               ; Read multiplication product (high byte).
+    phx : php
+    sep #$30
+
+    pha
+    inc
+    and $01,s
+    beq ?.powOf2                        ; max of the form 2^n - 1, so we can just use a bitwise and instead of a divide
+
+?.divide:
+    if !sa1 == 0
+        jsl $01ACF9|!bank               ; first byte in both A and $148C, second in $148D
+        sta $4204                       ; dividend, low byte
+        lda $148D|!addr : sta $4205     ; dividend, high byte
+        pla : inc
+        sta $4206                       ; divisor
+        nop #8                          ; wait 16 cycles
+        lda $4216                       ; remainder
     else
-        STZ $2250               ; Set multiplication mode.
-        STA $2251               ; Write first multiplicand.
-        STZ $2252
-        STX $2253               ; Write second multiplicand.
-        STZ $2254
-        NOP                     ; Wait 2 cycles, which is enough according to SnesLab docs about sa-1 registers.
-        LDA $2307               ; Read multiplication product.
+        lda #$01 : sta $2250            ; select division
+        jsl $01ACF9|!bank               ; first byte in both A and $148C, second in $148D
+        sta $2251                       ; dividend, low byte
+        lda $148D|!addr : lsr           ; workaround for snes9x bug: https://github.com/snes9xgit/snes9x/issues/799 (don't really want a negative dividend anyway)

```

Comment:
Do we really need to work around an emulator bug? This was fixed with snes9x 1.62, which has been released for a year at this point. If it is being used, then we're not at fault.

Reply:
True, but it's a very easy bug to work around, and it potentially saves hack creators the pain of having to track down why their boss is breaking for just some people with no idea why.  Keeping the dividend positive also isn't a terrible idea in this case anyway.

Reply:
Fair enough, I'm still annoyed at the idea of having to keep extra instructions around because of an emulator bug but wcyd